### PR TITLE
Unit test for pipeline exec engine

### PR DESCRIPTION
**Pull request by liuyehcf**
1. unit test for pipeline control flow

### CODE REVIEW

File: be/src/storage/rowset/rowset_factory.cpp [hunk 1]
```diff
@@ -36,7 +36,7 @@ OLAPStatus RowsetFactory::create_rowset(MemTracker* mem_tracker, const TabletSch
                                         const std::string& rowset_path, const RowsetMetaSharedPtr& rowset_meta,
                                         RowsetSharedPtr* rowset) {
     if (rowset_meta->rowset_type() == BETA_ROWSET) {
-        rowset->reset(new BetaRowset(mem_tracker, schema, rowset_path, rowset_meta));
+        *rowset = std::make_shared<BetaRowset>(mem_tracker, schema, rowset_path, rowset_meta);
```
Comment (Pslydhh) [inline be/src/storage/rowset/rowset_factory.cpp:39]: What is the difference between reset and assignment of std::shared_ptr in these cases?

File: be/src/exec/parquet/file_reader.cpp [hunk 2]
```diff
@@ -102,7 +102,8 @@ Status FileReader::_parse_footer() {
 
     tparquet::FileMetaData t_metadata;
     // deserialize footer
-    RETURN_IF_ERROR(deserialize_thrift_msg(footer_buf + to_read - 8 - footer_size, &footer_size, true, &t_metadata));
+    RETURN_IF_ERROR(deserialize_thrift_msg(footer_buf + to_read - 8 - footer_size, &footer_size, TProtocolType::COMPACT,
```
Comment (satanson) [inline be/src/exec/parquet/file_reader.cpp:105]: add some comment to explain the meaning of TProtocolType::COMPACT

File: be/src/exec/pipeline/analysis/analytic_sink_operator.h [hunk 3]
```diff
@@ -35,7 +35,7 @@ class AnalyticSinkOperator : public Operator {
 class AnalyticSinkOperatorFactory final : public OperatorFactory {
 public:
     AnalyticSinkOperatorFactory(int32_t id, int32_t plan_node_id, AnalytorPtr analytor)
-            : OperatorFactory(id, plan_node_id), _analytor(analytor) {}
+            : OperatorFactory(id, "analytor_sink", plan_node_id), _analytor(analytor) {}
```
Comment (satanson) [inline be/src/exec/pipeline/analysis/analytic_sink_operator.h:38]: typo "analytor_sink"

File: be/src/exec/pipeline/pipeline_driver_dispatcher.cpp [hunk 4]
```diff
@@ -45,7 +49,9 @@ void GlobalDriverDispatcher::run() {
 
         size_t queue_index;
         auto driver = this->_driver_queue->take(&queue_index);
-        DCHECK(driver != nullptr);
+        if (driver == nullptr) {
```
Comment (satanson) [inline be/src/exec/pipeline/pipeline_driver_dispatcher.cpp:52]: _driver_queue always return a non-null driver
"DCHECK(driver != nullptr);"  is ok

File: be/src/exec/pipeline/query_context.h [hunk 5]
```diff
@@ -78,14 +79,16 @@ class QueryContextManager {
...
     QueryContext* get_or_register(const TUniqueId& query_id);
     QueryContextPtr get(const TUniqueId& query_id);
+    QueryPromisePtr get_promise(const TUniqueId& query_id);
     QueryContextPtr remove(const TUniqueId& query_id);
 
 private:
     //TODO(by satanson)
     // A multi-shard map may be more efficient
     std::mutex _lock;
+    std::unordered_map<TUniqueId, QueryPromisePtr> _promises;
```
Comment (kangkaisen) [inline be/src/exec/pipeline/query_context.h:89]: Why do we need to add the `_promises`?

Comment (liuyehcf) [reply be/src/exec/pipeline/query_context.h:89]: same as above

File: be/src/exec/parquet/file_reader.cpp (hunk 2 above)
Comment (liuyehcf) [reply be/src/exec/parquet/file_reader.cpp:105]: Add comment to the definition of enum TProtocolType

File: be/src/exec/pipeline/query_context.h [hunk 6]
```diff
@@ -78,14 +79,16 @@ class QueryContextManager {
 public:
     QueryContext* get_or_register(const TUniqueId& query_id);
     QueryContextPtr get(const TUniqueId& query_id);
+    QueryPromisePtr get_promise(const TUniqueId& query_id);
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply be/src/exec/pipeline/query_context.h:82]: 1. The promise should reflect in function return type, not the function name. The promise expresses finishing semantics, which should reflect in the function name
2. The promise only be used the UT, but will effect the normal code path performance. Could we find a way don't effect the normal code path performance?

File: be/src/exec/pipeline/pipeline_driver_dispatcher.cpp [hunk 7]
```diff
@@ -44,8 +48,13 @@ void GlobalDriverDispatcher::run() {
         }
 
         size_t queue_index;
-        auto driver = this->_driver_queue->take(&queue_index);
+        auto take_driver = this->_driver_queue->take(&queue_index);
```
Comment (satanson) [inline be/src/exec/pipeline/pipeline_driver_dispatcher.cpp:51]: rename take_driver to maybe_driver/driver_status/driver_result.

**Review by satanson — APPROVED** [review]

**Review by Pslydhh — APPROVED** [review]

File: be/src/exec/pipeline/query_context.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (liuyehcf) [reply be/src/exec/pipeline/query_context.h]: fixed

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**